FEXCore/Source/Interface/Core/OpcodeDispatcher/AVX_128.cpp

```diff
@@ -617,4 +652,134 @@ void OpDispatchBuilder::AVX128_VZERO(OpcodeArgs) {
+    Ref ZeroVector = LoadZeroVector(OpSize::i128Bit);
+
+    AVX128_StoreResult_WithOpSize(Op, Op->Dest, RefPair {.Low = Result_Low, .High = ZeroVector});
+  } else {
+    // Need to store Bits[127:64]. Duplicate the element to get it in the low bits.
+    Src1.Low = _VDupElement(OpSize::i128Bit, OpSize::i64Bit, Src1.Low, 1);
+    StoreResult_WithOpSize(FPRClass, Op, Op->Dest, Src1.Low, OpSize::i64Bit, OpSize::i64Bit);
+  }
+}
+
+void OpDispatchBuilder::AVX128_VMOVDDUP(OpcodeArgs) {
+  const auto SrcSize = GetSrcSize(Op);
+  const auto Is128Bit = SrcSize == Core::CPUState::XMM_SSE_REG_SIZE;
+
+  const auto IsSrcGPR = Op->Src[0].IsGPR();
+
+  RefPair Src {};
+  if (IsSrcGPR) {
+    Src = AVX128_LoadSource_WithOpSize(Op, Op->Src[0], Op->Flags, !Is128Bit);
+  } else {
+    // Accesses from memory are a little weird.
+    // 128-bit operation only loads 8-bytes.
+    // 256-bit operation loads a full 32-bytes.
+    if (Is128Bit) {
+      Src.Low = LoadSource_WithOpSize(FPRClass, Op, Op->Src[0], OpSize::i128Bit, Op->Flags);
+    } else {
+      Src = AVX128_LoadSource_WithOpSize(Op, Op->Src[0], Op->Flags, !Is128Bit);
+    }
+  }
+
+  if (Is128Bit) {
+    // Duplicate Src[63:0] in to low 128-bits
+    auto Result_Low = _VDupElement(OpSize::i128Bit, OpSize::i64Bit, Src.Low, 0);
+    Ref ZeroVector = LoadZeroVector(OpSize::i128Bit);
+
+    AVX128_StoreResult_WithOpSize(Op, Op->Dest, RefPair {.Low = Result_Low, .High = ZeroVector});
+  } else {
+    // Duplicate Src.Low[63:0] in to low 128-bits
+    auto Result_Low = _VDupElement(OpSize::i128Bit, OpSize::i64Bit, Src.Low, 0);
+    // Duplicate Src.High[63:0] in to high 128-bits
+    auto Result_High = _VDupElement(OpSize::i128Bit, OpSize::i64Bit, Src.High, 0);
+    AVX128_StoreResult_WithOpSize(Op, Op->Dest, RefPair {.Low = Result_Low, .High = Result_High});
+  }
+}
+
+void OpDispatchBuilder::AVX128_VMOVSLDUP(OpcodeArgs) {
+  AVX128_VectorUnaryImpl(Op, GetSrcSize(Op), OpSize::i32Bit,
+                         [this](size_t ElementSize, Ref Src) { return _VTrn(OpSize::i128Bit, ElementSize, Src, Src); });
+}
+
+void OpDispatchBuilder::AVX128_VMOVSHDUP(OpcodeArgs) {
+  AVX128_VectorUnaryImpl(Op, GetSrcSize(Op), OpSize::i32Bit,
+                         [this](size_t ElementSize, Ref Src) { return _VTrn2(OpSize::i128Bit, ElementSize, Src, Src); });
+}
+
+template<size_t ElementSize>
+void OpDispatchBuilder::AVX128_VBROADCAST(OpcodeArgs) {
+  const auto DstSize = GetDstSize(Op);
+  const auto Is128Bit = DstSize == Core::CPUState::XMM_SSE_REG_SIZE;
+  RefPair Src {};
+
+  if (Op->Src[0].IsGPR()) {
+    Src = AVX128_LoadSource_WithOpSize(Op, Op->Src[0], Op->Flags, false);
+    if (ElementSize != OpSize::i128Bit) {
+      // Only duplicate if not VBROADCASTF128.
+      Src.Low = _VDupElement(OpSize::i128Bit, ElementSize, Src.Low, 0);
+    }
+  } else {
+    // Get the address to broadcast from into a GPR.
+    Ref Address = MakeSegmentAddress(Op, Op->Src[0], CTX->GetGPRSize());
```

Comment:
How can 64-bit AVX instructions have segment registers?

Reply:
FS and GS are still a thing for memory accesses.

Reply:
Oh. TLS. Kewl.